packages/platforms/accton/x86-64/x86-64-accton-as5912-54x/onlp/builds/src/module/src/fani.c

```diff
@@ -0,0 +1,369 @@
+/************************************************************
+ * <bsn.cl fy=2014 v=onl>
+ *
+ *           Copyright 2014 Big Switch Networks, Inc.
+ *           Copyright 2014 Accton Technology Corporation.
+ *
+ * Licensed under the Eclipse Public License, Version 1.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ *        http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the
+ * License.
+ *
+ * </bsn.cl>
+ ************************************************************
+ *
+ * Fan Platform Implementation Defaults.
+ *
+ ***********************************************************/
+#include <onlplib/file.h>
+#include <onlp/platformi/fani.h>
+#include "platform_lib.h"
+
+#define PSU_PREFIX_PATH  "/sys/bus/i2c/devices/"
+
+enum fan_id {
+	FAN_1_ON_FAN_BOARD = 1,
+	FAN_2_ON_FAN_BOARD,
+	FAN_3_ON_FAN_BOARD,
+	FAN_4_ON_FAN_BOARD,
+	FAN_5_ON_FAN_BOARD,
+	FAN_6_ON_FAN_BOARD,
+	FAN_1_ON_PSU_1,
+	FAN_1_ON_PSU_2,
+};
+
+#define MAX_FAN_SPEED     25500
+#define MAX_PSU_FAN_SPEED 25500
+
+#define CHASSIS_FAN_INFO(fid)		\
+    { \
+        { ONLP_FAN_ID_CREATE(FAN_##fid##_ON_FAN_BOARD), "Chassis Fan - "#fid, 0 },\
+        0x0,\
+        ONLP_FAN_CAPS_SET_PERCENTAGE | ONLP_FAN_CAPS_GET_RPM | ONLP_FAN_CAPS_GET_PERCENTAGE,\
+        0,\
+        0,\
+        ONLP_FAN_MODE_INVALID,\
+    }
+
+#define PSU_FAN_INFO(pid, fid) 		\
+    { \
+        { ONLP_FAN_ID_CREATE(FAN_##fid##_ON_PSU_##pid), "PSU "#pid" - Fan "#fid, 0 },\
+        0x0,\
+        ONLP_FAN_CAPS_SET_PERCENTAGE | ONLP_FAN_CAPS_GET_RPM | ONLP_FAN_CAPS_GET_PERCENTAGE,\
+        0,\
+        0,\
+        ONLP_FAN_MODE_INVALID,\
+    }
+
+/* Static fan information */
+onlp_fan_info_t finfo[] = {
+    { }, /* Not used */
+	CHASSIS_FAN_INFO(1),
+	CHASSIS_FAN_INFO(2),
+	CHASSIS_FAN_INFO(3),
+	CHASSIS_FAN_INFO(4),
+	CHASSIS_FAN_INFO(5),
+	CHASSIS_FAN_INFO(6),
+	PSU_FAN_INFO(1, 1),
+	PSU_FAN_INFO(2, 1)
+};
+
+#define VALIDATE(_id)                           \
+    do {                                        \
+        if(!ONLP_OID_IS_FAN(_id)) {             \
+            return ONLP_STATUS_E_INVALID;       \
+        }                                       \
+    } while(0)
+ 
+static int
+_onlp_fani_info_get_fan(int fid, onlp_fan_info_t* info)
+{
+	int   value;
+	char  path[64] = {0};
+
+	/* get fan present status
+	 */
+	sprintf(path, "%s""fan%d_present", FAN_BOARD_PATH, fid);
```

Comment:
@brandonchuang  This pattern happens so often that it is supported by the onlp_file_* APIs:
Please see the prototype here: https://github.com/opencomputeproject/OpenNetworkLinux/blob/master/packages/base/any/onlp/src/onlplib/module/src/file.c#L168

You can pass a format string and format arguments directly to the file  API like this:
``onlp_file_read_int(&value, "%s""fan%d_present", FAN_BOARD_PATH, fid)``

This will make your implementation easier.